File: 1239. Maximum Length of a Concatenated String with Unique Characters/1239. Maximum Length of a Concatenated String with Unique Characters.py
class Solution:
    def maxLength(self, arr: List[str]) -> int:
        def is_unique(s):
            return len(s) == len(set(s))
        
        def backtrack(index, current):
            if not is_unique(current):
                return 0
            
            max_length = len(current)
            
            for i in range(index, len(arr)):
                max_length = max(max_length, backtrack(i + 1, current + arr[i]))
            
            return max_length
        
        return backtrack(0, "")

File: 1239. Maximum Length of a Concatenated String with Unique Characters/1239. Maximum Length of a Concatenated String with Unique Characters.rs
impl Solution {
    pub fn max_length(arr: Vec<String>) -> i32 {
        fn is_unique(s: &str) -> bool {
            let mut chars = vec![false; 26];
            for ch in s.chars() {
                if chars[ch as usize - 'a' as usize] {
                    return false;
                }
                chars[ch as usize - 'a' as usize] = true;
            }
            true
        }

        fn backtrack(arr: &Vec<String>, index: usize, current: &str) -> i32 {
            if !is_unique(current) {
                return 0;
            }

            let mut max_length = current.len() as i32;

            for i in index..arr.len() {
                let mut new_current = current.to_string();
                new_current.push_str(&arr[i]);
                max_length = max_length.max(backtrack(arr, i + 1, &new_current));
            }

            max_length
        }

        backtrack(&arr, 0, "")
    }
}


